src/Console.php

```diff
@@ -0,0 +1,249 @@
+                $this->output('Error: '.$e->getMessage());
+            } catch (\Exception $e) {
+                $this->output('Exception: '.$e->getMessage());
+            }
+            $this->sseInProgress = false;
+        });
+        $this->js(true, $this->sse);
+
+        return $this;
+    }
+
+    /**
+     * Output a single line to the console.
+     *
+     * @param $text string
+     *
+     * @return $this
+     */
+    public function output($text)
+    {
+        $this->_output_bypass = true;
+        $this->sse->send($this->js()->append(htmlspecialchars($text).'<br/>'));
+        $this->_output_bypass = false;
+
+        return $this;
+    }
+
+    /**
+     * Output un-escaped HTML line. Use this to send HTML.
+     *
+     * @param $text string
+     *
+     * @return $this
+     */
+    public function outputHTML($text)
+    {
+        $this->_output_bypass = true;
+        $this->sse->send($this->js()->append($text.'<br/>'));
+        $this->_output_bypass = false;
+    }
+
+    /**
+     * Executes command passing along escaped arguments.
+     *
+     * Will also stream stdout / stderr as the comand executes.
+     * once command terminates method will return the exit code.
+     *
+     * This method can be executed from inside callback or
+     * without it.
+     */
+
+    /*
+    public function runCommand($exec, $args = [])
+    {
+        if (!$this->sseInProgress) {
+            $this->set(function () {
+                $this->runCommand($exec, $args);
+            });
+
+            return;
+        }
+
+
+        // not implemented here
+        //
+        //
+    }
+     */
+
+    /**
+     * Execute method of a certain model. That's a short-hand method
+     * for running:.
+     *
+     * $app->add('Console')->setModel(new User($db), 'generateReports');
+     *
+     * You can enable output from inside your method if you:
+     *
+     *  - implement \atk4\core\DebugTrait in your model
+     *  - use $this->debug() or $this->info()
+     *  - if you wish to get log from other objects, be sure to switch debug on with $obj->debug = true;
+     *
+     * @param $model \atk4\data\Model
+     * @param $method string
+     * @param $args array
+     */
+    public function setModel(\atk4\data\Model $model, string $method, $args = [])
+    {
+        if (!$this->sseInProgress) {
+            $this->set(function () use ($model, $method, $args) {
+                $this->setModel($model, $method, $args);
+            });
+
+            return;
+        }
+
+        // temporarily override app logging
+        if (isset($model->app)) {
+            $old_logger = $model->app->logger;
+            $model->app->logger = $this;
+        }
+
+        $this->output('--[ Executing '.get_class($model).'->'.$method.' ]--------------');
+        $model->debug = true;
+        $result = call_user_func_array([$model, $method], $args);
+        $this->output('--[ Result: '.json_encode($result).' ]------------');
```

Comment:
why json_encode here? goal is to convert anything callback method returns to string, right?

Reply:
well, something better may be there, but not sure what. either that or export, and i dont like var_export.